## ¿Existe relación alguna entre el total de minutos usados con la compañía y el ingreso total de los clientes?
En el desarrollo de este caso introduciremos las librerías pandas, numpy y matplotlib para el preprocesamiento, visualización de un conjunto de datos y el desarrollo de algunas medidas estadísticas descriptivas.



##  Introducción 

La revisión del comportamiento de los clientes en cualquier empresa es de vital importancia para el desarrollo de estrategias  tanto de mejoramiento de productos como de publicidad y toma de decisiones. Evaluaremos la información obtenida de un conjunto de registro obtenidos de una empresa de telecomunicaciones que cuenta con la revisión de las siguientes variables:

1. totrev: Ingresos del cliente
2. totmou: Total de minutos usados por el cliente
3. area: Área geográfica
4. creditcd: Indicador de tarjeta de crédito
5. eqpdays: Número de días (antigüedad) del equipo actual

Es importante mencionar que la empresa contiene una gran cantidad de clientes y hacer el análisis uno a uno no es lo  óptimo. Con ayuda de la programación y el análisis de datos, se logrará llegar a conclusiones rápidas y acertadas para encontrar relaciones que a simple vista no son tenidas en cuenta.

### ¿Qué se quiere identificar? 

Teniendo en cuenta la información de la base de datos se quiere identificar si hay una relación entre el total de minutos usados por el cliente con la compañía y el ingreso total de los clientes, es decir, Será posible afirmar que cuando una persona
tiene un mayor ingreso mayor entonces es mayor la utilización de minutos o todo lo contrario?


####  Insumos para tratar este problema


Teniendo en cuenta que la compañía tenia estándares para la entrega de la información, se dispone de un  conjunto de tablas en formato CSV ordenado de la siguiente manera:
La información de este conjunto de datos puede encontrase en [aquí](https://www.kaggle.com/abhinav89/telecom-customer?select=Telecom_customer+churn.csv)

### Objetivos

En este caso usted tendrá que cargar varias bases de datos, hará una exploración básica sobre la información y fusionará las distintas bases para tener una visión general del problema. La idea es que desarrolle las siguientes habilidades:

1. Manejo de la librería pandas de Python para cargar y leer datos; 
2. Ideas elementales para una útil transformación de los datos;
3. Construcción y presentación de argumentos válidos que le brindarán una solución para la pregunta expresada más arriba y llegar a una conclusión. 
4. Visualización de gráficas elementales que le permitirá entender mejor la información contenida en los datos.

#### Importación de paquetes


Una de las mejores opciones para trabajar con datos tabulares en Python es usar el módulo pandas. La librería `pandas` provee estructuras de datos, genera gráficos de alta calidad con `matplotlib` y se integra de buena forma con otras librerías que usan arrays de `numpy`.

Debemos revisar si la librería se encuentra en nuestro sistema usando el comando 

```python
!pip show librería
```


In [1]:
#!pip show pandas
#!pip show numpy
#!pip show matplotlib

En el caso de que la librería no se encuentre instalada ejecutamos el comando

```python
!pip install librería
```

Para importar las librerías usamos el comando 
```python
import librería as alias
```

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

Luego de que tener los paquetes activados dentro de nuestro ambiente de python podremos usar las fucnciones que se encuentren en la librería.

### Introducción a pandas

Poner información sobre pandas e indicar la diferencia entre series y dataframe

![](https://www.cdn.geeksforgeeks.org/wp-content/uploads/creating_dataframe1.png)

A continuación, mostramos el código a ejecutar para crear una ` Serie`

In [3]:
Serie1 = pd.Series(
index=['Primero', 'Segundo', 'Tercero', 'Cuarto', 'Quinto'], name="Variable1", data=["A","B","C","D","E"]
)
Serie1

Primero    A
Segundo    B
Tercero    C
Cuarto     D
Quinto     E
Name: Variable1, dtype: object

In [4]:
Data_Frame = pd.DataFrame( 
index=['Primero', 'Segundo', 'Tercero', 'Cuarto', 'Quinto'],
    columns=["Variable2","Variable3"], 
    data=[['Hola', 12], [2, 22],[3,32],[4,42],[5,52]]
)
Data_Frame

Variable2  Variable3
Primero      Hola         12
Segundo         2         22
Tercero         3         32
Cuarto          4         42
Quinto          5         52

Para revisar el tipo de objeto que contiene nuestro DataFrame podemos ejecutar

In [5]:
Data_Frame.dtypes
#Serie1.dtypes

Variable2    object
Variable3     int64
dtype: object

Extraer datos sobre la base:

Si quiero conocer una columna de la base simplemente escribo `DataFrame['nombre_columna']`

In [6]:
Data_Frame[['Variable2']]

Variable2
Primero      Hola
Segundo         2
Tercero         3
Cuarto          4
Quinto          5

In [7]:
Data_Frame['Variable1']=Serie1

In [8]:
Data_Frame

Variable2  Variable3 Variable1
Primero      Hola         12         A
Segundo         2         22         B
Tercero         3         32         C
Cuarto          4         42         D
Quinto          5         52         E

Si quiero ver varias columnas de la base de datos, utilizo doble paréntesis cuadrado:

`DataFrame[['columna1','columna2']]`

In [9]:
Data_Frame[['Variable1','Variable2']]

Variable1 Variable2
Primero         A      Hola
Segundo         B         2
Tercero         C         3
Cuarto          D         4
Quinto          E         5

Si quiero ver datos por registros utilizo loc:

In [10]:
Data_Frame.loc[['Primero','Tercero','Quinto']]

Variable2  Variable3 Variable1
Primero      Hola         12         A
Tercero         3         32         C
Quinto          5         52         E

In [11]:
Data_Frame.iloc[4]

Variable2     5
Variable3    52
Variable1     E
Name: Quinto, dtype: object

### Carga de información en `pandas`

Para nuestro  ejercicio usaremos los siguientes conjuntos de datos:

1. NEW ENGLAND AREA.csv
2. ATLANTIC SOUTH AREA.csv
3. NEW YORK CITY AREA.csv
4. CHICAGO AREA.csv
5. NORTH FLORIDA AREA.csv
6. DC-MARYLAND-VIRGINIA AREA.csv
7. GREAT LAKES AREA.csv
8. NORTHWEST-ROCKY MOUNTAIN AREA.csv
9. MIDWEST AREA.csv
10. SOUTHWEST AREA.csv
11. LOS ANGELES AREA.csv
12. HOUSTON AREA.csv
13. CALIFORNIA NORTH AREA.csv
14. CENTRAL-SOUTH TEXAS AREA.csv
15. DALLAS AREA.csv
16. PHILADELPHIA AREA.csv
17. TENNESSEE AREA.csv
18. OHIO AREA.csv
19. SOUTH FLORIDA AREA.csv

Cada uno de estos conjuntos de datos contiene la información de:

1. *mou_Mean:* Número medio de minutos de uso mensuales
2. *custcare_Mean:* Número medio de llamadas de atención al cliente
3. *area:*  Área geografica
4. *creditcd:* Indicador de tarjeta de crédito
5. *eqpdays:* Número de días (antigüedad) del equipo actual

Además, se incluye _Credito_dias.csv_ con la información de 2 variables adicionales para los registros las cuales son:

1. _creditcd:_ Indicador de tarjeta de crédito
2. _eqpdays:_ Número de días (antigüedad) del equipo actual

Revisaremos inicialmente los datos contenidos en _Credito_dias.csv_ para revisar algunas funciones interesantes de `pandas`, usaremos el comando 
```python
pd.read_csv("archivo.csv")
```
Podemos definir el número de registros a imprimir usando
```python
pd.options.display.min_rows=10
````

In [12]:
df=pd.read_csv("Bases/Credito_dias.csv", sep=",",index_col="Customer_ID")
df

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1076695       4155.0   433.16        N      222
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[10000 rows x 4 columns]

El conjunto de datos se encuentra en el `DataFrame` `df`

sobre el objeto creado podemos usar  algunas funciones  para darnos una idea del comportamiento de la información:
```python
df.head(n)  # imprime los primeros n registros del DataFrame
df.tail(n)  # imprime los últimos n registros del DataFrame 
df.shape    # imprime el numero de columnas y filas del DataFrame
df.columns  # imprime el nombre de las columnas del DataFrame
df.index    # imprime el indice de los registros del DataFrame
df.dtypes   # imprime el tipo de cada una de las columnnas del DataFrame
df.sample(n)# imprime una muestra aleatoria de n registros en el DataFrame
```

In [13]:
display("primeros 3 registros",df.head(3))
display("3 últimos registros",df.tail(3))
display("Columnas:",df.columns)
display("Índices:",df.index)
display("Tipos de registros en el DataFrame",df.dtypes)
display("Muestra aleatoria de 3 registros",df.sample(3))
display("dimensión",df.shape)

'primeros 3 registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675

'3 últimos registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                   
1096497      3690.0   294.29        Y      212
1087248      3333.0   233.08        Y      290
1099003      5025.0  1436.49        Y     1029

'Columnas:'

Index(['totmou', 'totrev', 'creditcd', 'eqpdays'], dtype='object')

'Índices:'

Index([1082423, 1069232, 1031705, 1089806, 1076695, 1091652, 1043846, 1088967,
       1045239, 1046584,
       ...
       1087145, 1002540, 1044082, 1083458, 1093976, 1027614, 1098422, 1096497,
       1087248, 1099003],
      dtype='int64', name='Customer_ID', length=10000)

'Tipos de registros en el DataFrame'

totmou      float64
totrev      float64
creditcd     object
eqpdays       int64
dtype: object

'Muestra aleatoria de 3 registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1043539      12151.0  1287.32        Y      308
1040020      16527.0  1111.81        Y      594
1007908         29.0   545.56        Y      986

'dimensión'

(10000, 4)

Este conjunto de datos corresponde a 10000 registros y 2 variables

### Manejo de datos duplicados 

os dataframes tienen la posibilidad de detectar las filas duplicadas con la función: 
```
df.duplicated({columns})
```
Si no especifica {columns}, se busacarán duplicados teniendo en cuenta todas las columnas. Una vez detectados el dtaframe tiene un función para eliminar filas duplicadas, haciendo:

```
df = df.drop_duplicates()
```
También es posible eliminar filas que duplican solo algunos campos. Con el fin de dejar solo una ocurrencia.

```
df.drop_duplicates(['nombre', 'apellido', 'cédula'], keep='last')
```
En este caso, como no todas las colomnas son iguales se conserva solo la última ocurrencia.

### Manejo de datos perdidos o esperados
Son aquellos datos flatantes en la tabla que por alguna razón o error no se encuentran, o parecen con algún valor o etiqueta de no válido. 
Para manipular datos perdidos use la explicación dada el siguiente enlace: [DATOS PERDIDOS](http://ligdigonzalez.com/manipulando-datos-perdidos-en-python/)
recuerde que para reemplazar datos perdidos en un dataframe $df$ se usa la función:
```python
promedio = df['totrev'].mean()
df['nota'].replace(np.nan, promedio)
```
en donde la opción 'promedio' busca el valor promedio de la columna (datos numéricos) o la moda (datos categóricos) para reemplazar el valor no válido.


In [14]:
df.isnull().any(axis=0)

totmou      False
totrev      False
creditcd     True
eqpdays     False
dtype: bool

In [15]:
perd=df.isnull().any(axis=1)

In [16]:
df[perd]

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1072218       5684.0   857.65      NaN      244
1066597        925.0   373.12      NaN      344
1023352      14466.0  2243.76      NaN      387
1045521       2504.0   756.79      NaN      495
1001813      76260.0  5526.11      NaN      361
...              ...      ...      ...      ...
1050795      15591.0  1330.71      NaN      447
1096985       3011.0   522.14      NaN       74
1025203        853.0   650.34      NaN      780
1099258      33660.0  4247.85      NaN      206
1026149       1955.0   606.08      NaN      770

[152 rows x 4 columns]

In [17]:
perdidos = df[df.isnull().any(axis=1)]
perdidos

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1072218       5684.0   857.65      NaN      244
1066597        925.0   373.12      NaN      344
1023352      14466.0  2243.76      NaN      387
1045521       2504.0   756.79      NaN      495
1001813      76260.0  5526.11      NaN      361
...              ...      ...      ...      ...
1050795      15591.0  1330.71      NaN      447
1096985       3011.0   522.14      NaN       74
1025203        853.0   650.34      NaN      780
1099258      33660.0  4247.85      NaN      206
1026149       1955.0   606.08      NaN      770

[152 rows x 4 columns]

In [18]:
df['totrev']

Customer_ID
1082423     906.58
1069232    1172.25
1031705     577.39
1089806     294.26
1076695     433.16
            ...   
1027614     484.31
1098422     583.36
1096497     294.29
1087248     233.08
1099003    1436.49
Name: totrev, Length: 10000, dtype: float64

In [19]:
promedio = df['totrev'].mean()
df['totrev']=df['totrev'].replace(np.nan, promedio)

También es recomendado usar la funcion `fillna()`

In [20]:
df['creditcd'].fillna("Y",inplace=True)

In [21]:
df[perd]

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1072218       5684.0   857.65        Y      244
1066597        925.0   373.12        Y      344
1023352      14466.0  2243.76        Y      387
1045521       2504.0   756.79        Y      495
1001813      76260.0  5526.11        Y      361
...              ...      ...      ...      ...
1050795      15591.0  1330.71        Y      447
1096985       3011.0   522.14        Y       74
1025203        853.0   650.34        Y      780
1099258      33660.0  4247.85        Y      206
1026149       1955.0   606.08        Y      770

[152 rows x 4 columns]

In [22]:
df.isnull().any(axis=0)

totmou      False
totrev      False
creditcd    False
eqpdays     False
dtype: bool

In [23]:
df

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1076695       4155.0   433.16        N      222
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[10000 rows x 4 columns]



### Cálculo de estadísticas resumen

Las medidas estadísticas resumen son de gran ayuda para comprender como es el comportamiento de la distribución de los datos, los objetos `DataFrame` ofrecen facilidades para hacer el cálculo de algunas estadísticas sobre cada una de las variables.

```python
df["var"].value_counts() # Encuentra la frecuencia de las categorias de una columna (la columna debe ser cualitativa)
df["var"].min()          # Encuentra el mínimo de una variable cuantitativa
df["var"].median()       # Encuentra el mínimo de una variable cuantitativa
df["var"].mean()         # Encuentra el promedio de una variable cuantitativa
df["var"].max()          # Encuentra el máximo de una variable cuantitativa
df["var"].quantile()     # Encuentra los cuantiles de una variable cuantitativa
```


In [24]:
df["creditcd"].value_counts()

creditcd
Y    6948
N    3052
Name: count, dtype: int64

In [25]:
df["creditcd"].count()

10000

In [26]:
df["eqpdays"].min()

-5

In [27]:
df["eqpdays"]=[0 if i<0 else i for i in df["eqpdays"] ]

In [28]:
df["eqpdays"].min()

0

La exploración del anterior resultado puede mostrar inconsistencias  en los registros y nos ayudará a tomar decisiones sobre algunos registros, el valor mínimo de la variable `eqpdays` (Número de días (antigüedad) del equipo actual) debería ser estrictamente positivo y estos pueden ser errores de ingreso.

In [29]:
df["eqpdays"].max()

1823

In [30]:
df["eqpdays"].median()

338.0

In [31]:
df["eqpdays"].mean()

386.7992

In [32]:
df["eqpdays"].quantile(0.25)

210.0

In [33]:
df["eqpdays"].quantile(0.75)

522.0

Estas estadísticas pueden ser resumidas con el comando:
```python
df["var"].describe()  # Estadísticas resumen principales
```

In [34]:
df["eqpdays"].describe()

count    10000.00000
mean       386.79920
std        252.71297
min          0.00000
25%        210.00000
50%        338.00000
75%        522.00000
max       1823.00000
Name: eqpdays, dtype: float64

In [35]:
df['creditcd'].describe()

count     10000
unique        2
top           Y
freq       6948
Name: creditcd, dtype: object

In [36]:
df.describe()

totmou        totrev      eqpdays
count   10000.000000  10000.000000  10000.00000
mean     7692.804866   1042.133807    386.79920
std      8657.917883    857.144895    252.71297
min         0.000000     11.100000      0.00000
25%      2576.000000    526.787500    210.00000
50%      5262.000000    820.235000    338.00000
75%      9823.250000   1272.975000    522.00000
max    163264.176700  13426.600000   1823.00000

In [37]:
df.describe(include='all')

totmou        totrev creditcd      eqpdays
count    10000.000000  10000.000000    10000  10000.00000
unique            NaN           NaN        2          NaN
top               NaN           NaN        Y          NaN
freq              NaN           NaN     6948          NaN
mean      7692.804866   1042.133807      NaN    386.79920
std       8657.917883    857.144895      NaN    252.71297
min          0.000000     11.100000      NaN      0.00000
25%       2576.000000    526.787500      NaN    210.00000
50%       5262.000000    820.235000      NaN    338.00000
75%       9823.250000   1272.975000      NaN    522.00000
max     163264.176700  13426.600000      NaN   1823.00000

### Agregar información de múltiples tablas

Hemos desarrollado un breve resumen estadístico solamente usando la base `Credito_dias.csv`. Realizaremos la combinación de las 19 áreas que nos reportan en los archivos csv. Una forma de lograr esta tarea de agregación es usar el método pd.concat() de pandas. Una entrada en este método puede ser una lista de DataFrames que quiera concatenar. Usaremos un ciclo  `for` sobre cada uno de los archivos de áreas para 


In [38]:
df.columns
df.dtypes

totmou      float64
totrev      float64
creditcd     object
eqpdays       int64
dtype: object

### Ejercicio 1
Determinar los percentiles 25, 50 y 75 para las columnas ttomou, totrev, eqpdays

In [39]:
#Espacio para el desarrollo del ejercicio
df["totmou"].describe()

count     10000.000000
mean       7692.804866
std        8657.917883
min           0.000000
25%        2576.000000
50%        5262.000000
75%        9823.250000
max      163264.176700
Name: totmou, dtype: float64

In [40]:
df["totrev"].describe()

count    10000.000000
mean      1042.133807
std        857.144895
min         11.100000
25%        526.787500
50%        820.235000
75%       1272.975000
max      13426.600000
Name: totrev, dtype: float64

In [41]:
df.describe(include='all')

totmou        totrev creditcd      eqpdays
count    10000.000000  10000.000000    10000  10000.00000
unique            NaN           NaN        2          NaN
top               NaN           NaN        Y          NaN
freq              NaN           NaN     6948          NaN
mean      7692.804866   1042.133807      NaN    386.79920
std       8657.917883    857.144895      NaN    252.71297
min          0.000000     11.100000      NaN      0.00000
25%       2576.000000    526.787500      NaN    210.00000
50%       5262.000000    820.235000      NaN    338.00000
75%       9823.250000   1272.975000      NaN    522.00000
max     163264.176700  13426.600000      NaN   1823.00000

In [42]:
os.listdir("Bases")

['.ipynb_checkpoints',
 'ATLANTIC SOUTH AREA.csv',
 'CALIFORNIA NORTH AREA.csv',
 'CENTRAL-SOUTH TEXAS AREA.csv',
 'CHICAGO AREA.csv',
 'Credito_dias.csv',
 'DALLAS AREA.csv',
 'DC-MARYLAND-VIRGINIA AREA.csv',
 'GREAT LAKES AREA.csv',
 'HOUSTON AREA.csv',
 'LOS ANGELES AREA.csv',
 'MIDWEST AREA.csv',
 'NEW ENGLAND AREA.csv',
 'NEW YORK CITY AREA.csv',
 'NORTH FLORIDA AREA.csv',
 'NORTHWEST-ROCKY MOUNTAIN AREA.csv',
 'OHIO AREA.csv',
 'PHILADELPHIA AREA.csv',
 'SOUTH FLORIDA AREA.csv',
 'SOUTHWEST AREA.csv',
 'TENNESSEE AREA.csv']

In [43]:
print("Definición de los símbolos de las acciones")
areas = ['ATLANTIC SOUTH AREA',
 'CALIFORNIA NORTH AREA',
 'CENTRAL-SOUTH TEXAS AREA',
 'CHICAGO AREA',
 'DALLAS AREA',
 'DC-MARYLAND-VIRGINIA AREA',
 'GREAT LAKES AREA',
 'HOUSTON AREA',
 'LOS ANGELES AREA',
 'MIDWEST AREA',
 'NEW ENGLAND AREA',
 'NEW YORK CITY AREA',
 'NORTH FLORIDA AREA',
 'NORTHWEST-ROCKY MOUNTAIN AREA',
 'OHIO AREA',
 'PHILADELPHIA AREA',
 'SOUTH FLORIDA AREA',
 'SOUTHWEST AREA',
 'TENNESSEE AREA']
lista_de_df = []
# Bucle sobre los símbolos
print(" --- Inicie el bucle sobre los símbolos --- ")
for i in areas:
    print("Procesando el símbolo: " + i)
    temp_df = pd.read_csv("Bases/" + i + ".csv",index_col="Customer_ID")
    temp_df["area"] = i
    lista_de_df.append(temp_df)
    # Usando un salto de línea al final de esta cadena de caracteres por estética
    print(" --- Bucle completo sobre los símbolos --- \n")
    # Combinando en un solo DataFrame usando el concat
print("Agregando los datos")
agr_df = pd.concat(lista_de_df, axis=0)
print(agr_df.shape)
print("Cabeza del DataFrame agr_df: ")
agr_df.head()

Definición de los símbolos de las acciones
 --- Inicie el bucle sobre los símbolos --- 
Procesando el símbolo: ATLANTIC SOUTH AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: CALIFORNIA NORTH AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: CENTRAL-SOUTH TEXAS AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: CHICAGO AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: DALLAS AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: DC-MARYLAND-VIRGINIA AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: GREAT LAKES AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: HOUSTON AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: LOS ANGELES AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: MIDWEST AREA
 --- Bucle completo sobre los símbolos --- 

Procesando el símbolo: NEW ENGLAND AREA
 --- Bucle com

custcare_Mean new_cell                 area creditcd
Customer_ID                                                      
1069232           3.666667        U  ATLANTIC SOUTH AREA        N
1043846           0.000000        U  ATLANTIC SOUTH AREA        Y
1008539           0.000000        N  ATLANTIC SOUTH AREA        Y
1068315           0.000000        U  ATLANTIC SOUTH AREA        N
1074220           1.666667        U  ATLANTIC SOUTH AREA        N

In [44]:
agr_df.sort_index(axis=0,inplace=True)
agr_df

custcare_Mean new_cell                           area creditcd
Customer_ID                                                                
1000014           0.000000        Y       CENTRAL-SOUTH TEXAS AREA        N
1000016           1.000000        N                    DALLAS AREA        Y
1000021           0.333333        U                 TENNESSEE AREA        Y
1000023           0.000000        N  NORTHWEST-ROCKY MOUNTAIN AREA        N
1000029           0.000000        N               NEW ENGLAND AREA        Y
...                    ...      ...                            ...      ...
1099907          17.333333        U               LOS ANGELES AREA        Y
1099932           0.000000        U               LOS ANGELES AREA        Y
1099963           0.000000        U                 SOUTHWEST AREA        Y
1099981           0.333333        U                 SOUTHWEST AREA      NaN
1099992           0.000000        Y               LOS ANGELES AREA        Y

[9997 rows x 4 columns]

La ejecución nos da como resultado una base de datos de 9997 registros correspondiente a la concatenación de todas las áreas. Se debe notar que nuestro conjunto de datos también es indexado por la columna `Customer_ID` y que la dimensión de los registros de está nueva base es menor a la de  `Creditos_dias.csv`.

 ## Cruce de información

El cruce de información es una operación esencial en el análisis de los datos. Usualmente, la posibilidad de realizar el cruce de dos o más fuentes de información es el primer paso para responder preguntas complejas acerca de los datos. Existen muchas formas de cruzar o combinar tablas de datos. El cruce de estos datos siempre tiene un propósito. Por tanto, se debe realizar un análisis de lo que se requiere antes de iniciar. Para esto, lo primero es conocer bien las fuentes de información que se van a relacionar. 

### Uniones (Joins)
Usando la librería pandas se pueden realizar 4 tipos de uniones para cruzar información entre dos fuentes de datos, ver Figura abajo. 
* INNER JOIN: este tipo de cruce incluye solo los resgistros que coincieden en las dos tablas
* LEFT JOIN:  en este tipo de cruce se incluyen todos los registros de la primera tabla aunque no aparezcan coincidencias en la segunda. Cuando no hay coincidencias en la segunda tabla los campos correspondientes a esta aparecerán como nulos.
* RIGHT JOIN: en este tipo de cruce se incluyen todos los registros de la segunda tabla aunque no aparezcan coincidencias en la primera. Cuando no hay coincidencias en la primer tabla los campos correspondientes a esta aparecerán como nulos.
* OUTER JOIN O FULL JOIN: combina todos los registros de ambas tablas aunque no existan coincidencias en la otra tabla. 

![Los tipos de uniones que se pueden ejecutar en pandas](https://letsdobigdata.files.wordpress.com/2016/03/joins.png)

La combinación a usar depende del análisis que se requiera con los datos. Lo más común es realizar un cruce de tipo Inner Join, debido a que muestra los registros cuando aparecen relacionados en ambas tablas. Sin embargo, en 
ocasiones se quiere saber que pasa con todos los registros de la primera tabla aunque no se tengan coincidencias. Alternativamente, quizá, lo menos usual es hacer combinaciones de tipo RIGHT o OUTER. Generalmente, estas son usadas cuando se quiere saber por qué no se tienen coincidencias, o cuáles registros no cumplen con la relación especificada.

Ya que desconocemos cuales son los registros perdidos en los 2 dataframes que hemos construido y es necesario para continuar nuestro preprocesamiento el unir la información.

In [45]:
df3=df.merge(agr_df,left_index=True,right_index=True,how="inner")
df3

totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1082423       7269.0   906.58          Y      371      14.333333        U   
1069232      13552.0  1172.25          N       83       3.666667        U   
1031705       4152.0   577.39          Y      675       3.000000        U   
1089806       1863.0   294.26          Y      218      11.000000        U   
1076695       4155.0   433.16          N      222       0.000000        U   
...              ...      ...        ...      ...            ...      ...   
1027614       2570.0   484.31          Y      611       0.666667        U   
1098422       2105.0   583.36          Y     1243       0.000000        U   
1096497       3690.0   294.29          Y      212       0.666667        Y   
1087248       3333.0   233.08          Y      290       0.000000        U   
1099003       5025.0  1436.49          Y     1029       0.333333        U   

                              area creditcd_y  
Customer_ID                                    
1082423           NEW ENGLAND AREA          Y  
1069232        ATLANTIC SOUTH AREA          N  
1031705         NEW YORK CITY AREA          Y  
1089806               CHICAGO AREA          Y  
1076695         NORTH FLORIDA AREA          N  
...                            ...        ...  
1027614               CHICAGO AREA          Y  
1098422             SOUTHWEST AREA          Y  
1096497           NEW ENGLAND AREA          Y  
1087248      CALIFORNIA NORTH AREA          Y  
1099003           LOS ANGELES AREA          Y  

[9997 rows x 8 columns]

### Ejercicio 2  
¿Cuáles son los 3 registros que no se incluyen en df3 ?


In [46]:
df.index.isin(df3.index)

array([ True,  True,  True, ...,  True,  True,  True])

In [47]:
sin_registros=df.index.isin(df3.index)
df.iloc[~sin_registros,]

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1050270       1409.0   407.02        Y      102
1020629      15284.0  2370.62        N      445
1020451       3604.0   662.42        Y      675

Si deseamos realizar una filtración sobre algún valor relevante por ejemplo el área podremos revertir el proceso usando el operador `==`  que retorna `True` en el caso que dos objetos tengan el mismo valor

In [48]:
new_england=df3[df3["area"]=="NEW ENGLAND AREA"]
new_england


totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1082423       7269.0   906.58          Y      371      14.333333        U   
1076771        288.0   374.98          Y      284       0.000000        U   
1083737       1928.0   396.78          Y      261       0.000000        U   
1078557       2781.0   590.26          Y      276       0.000000        N   
1018015      24069.0  2129.74          Y      199       2.000000        N   
...              ...      ...        ...      ...            ...      ...   
1059454        942.0   625.98          Y      323       0.000000        U   
1036973       2538.0   981.41          Y      688       0.000000        U   
1073829       2812.0   454.64          Y      342       0.000000        U   
1083458      10628.0   965.57          Y      262       6.666667        U   
1096497       3690.0   294.29          Y      212       0.666667        Y   

                         area creditcd_y  
Customer_ID                               
1082423      NEW ENGLAND AREA          Y  
1076771      NEW ENGLAND AREA          Y  
1083737      NEW ENGLAND AREA          Y  
1078557      NEW ENGLAND AREA          Y  
1018015      NEW ENGLAND AREA          Y  
...                       ...        ...  
1059454      NEW ENGLAND AREA          Y  
1036973      NEW ENGLAND AREA          Y  
1073829      NEW ENGLAND AREA          Y  
1083458      NEW ENGLAND AREA          Y  
1096497      NEW ENGLAND AREA          Y  

[525 rows x 8 columns]

El comando   `df3["area"]=="NEW ENGLAND AREA"`

In [49]:
df3["area"]=="NEW ENGLAND AREA"

Customer_ID
1082423     True
1069232    False
1031705    False
1089806    False
1076695    False
           ...  
1027614    False
1098422    False
1096497     True
1087248    False
1099003    False
Name: area, Length: 9997, dtype: bool

devuelve una serie boleana  con el mismo número de filas de la base`df3`, donde cada valor es `True` o `False` dependiendo si el valor de `area` de un registro  toma el valor de "NEW ENGLAND AREA" 

### Agrupación por área geográfica

En el Analisis de la información se suelen usar funciones de resumen, las cuales plantean el agrupamiento de ciertas características por categorías. Para más información consultar [gropby().](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html).

Para nuestro desarrollo es de interés el revisar si el comportamiento de las variables se mantiene en cada una de las áreas geográficas.


In [50]:
df3.groupby("area")

Aquí, el objeto DataFrameGroupBy  contiene un objeto DataFrame para cada grupo (en este caso, un objeto DataFrame para cada símbolo). Específicamente, cada elemento del objeto es una tupla que contiene el identificador de grupo (en este caso el símbolo), y las filas correspondientes del DataFrame que tienen ese símbolo).

La librería `pandas`  permite iterar sobre el objeto groupby() para ver lo que hay dentro:

In [51]:
group_df = df3.groupby(["area"]) # Datos del grupo en df3 filtrados por el área
# Haciendo un bucle a través de los grupos
for item in group_df:
    print(type(item)) # Mostrando el tipo de artículo en df3
    print(item[0]) # área
    print(item[1].head()) # DataFrame

<class 'tuple'>
('ATLANTIC SOUTH AREA',)
              totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1069232      13552.0  1172.25          N       83       3.666667        U   
1043846       7294.0  1495.18          Y      460       0.000000        U   
1008539       4078.0  1257.05          Y     1019       0.000000        N   
1068315       5124.0   652.27          N      377       0.000000        U   
1074220       6648.0   590.21          N      295       1.666667        U   

                            area creditcd_y  
Customer_ID                                  
1069232      ATLANTIC SOUTH AREA          N  
1043846      ATLANTIC SOUTH AREA          Y  
1008539      ATLANTIC SOUTH AREA          Y  
1068315      ATLANTIC SOUTH AREA          N  
1074220      ATLANTIC SOUTH AREA          N  
<class 'tuple'>
('CALIFORNIA NORTH AREA',)
              totmou   totrev creditcd_x  eqpdays  cus

Usando el método  `groupby()` con el método `describe()` y apliquémoslo a cada símbolo para
analizar la distribución de las características la variable ----.

In [52]:
group_df = df3.groupby(["area"]) # Datos del grupo en df3 filtrados por el área
# Bucle a través de los grupos
for item in group_df:
    print("------Área: ", item[0])
    group_df1 = item[1]
    df_relevante = group_df[["totrev"]]
    print(df_relevante.describe())

------Área:  ('ATLANTIC SOUTH AREA',)
                               totrev                                    \
                                count         mean          std     min   
area                                                                      
ATLANTIC SOUTH AREA             592.0   903.497399   697.197572  106.95   
CALIFORNIA NORTH AREA           599.0  1136.348013   922.069148  105.14   
CENTRAL-SOUTH TEXAS AREA        409.0  1174.366577   920.804942   48.53   
CHICAGO AREA                    496.0   892.133206   669.068444   59.90   
DALLAS AREA                     518.0  1258.508784  1267.499852   84.59   
DC-MARYLAND-VIRGINIA AREA       640.0   941.692188   634.132033   86.78   
GREAT LAKES AREA                488.0   987.714324   733.521417  126.56   
HOUSTON AREA                    458.0  1127.294127   999.987571   92.38   
LOS ANGELES AREA                671.0  1031.714382   738.616227   96.74   
MIDWEST AREA                    685.0  1095.678701   945.25186

                               totrev                                    \
                                count         mean          std     min   
area                                                                      
ATLANTIC SOUTH AREA             592.0   903.497399   697.197572  106.95   
CALIFORNIA NORTH AREA           599.0  1136.348013   922.069148  105.14   
CENTRAL-SOUTH TEXAS AREA        409.0  1174.366577   920.804942   48.53   
CHICAGO AREA                    496.0   892.133206   669.068444   59.90   
DALLAS AREA                     518.0  1258.508784  1267.499852   84.59   
DC-MARYLAND-VIRGINIA AREA       640.0   941.692188   634.132033   86.78   
GREAT LAKES AREA                488.0   987.714324   733.521417  126.56   
HOUSTON AREA                    458.0  1127.294127   999.987571   92.38   
LOS ANGELES AREA                671.0  1031.714382   738.616227   96.74   
MIDWEST AREA                    685.0  1095.678701   945.251868   90.42   
NEW ENGLAND AREA         

                               totrev                                    \
                                count         mean          std     min   
area                                                                      
ATLANTIC SOUTH AREA             592.0   903.497399   697.197572  106.95   
CALIFORNIA NORTH AREA           599.0  1136.348013   922.069148  105.14   
CENTRAL-SOUTH TEXAS AREA        409.0  1174.366577   920.804942   48.53   
CHICAGO AREA                    496.0   892.133206   669.068444   59.90   
DALLAS AREA                     518.0  1258.508784  1267.499852   84.59   
DC-MARYLAND-VIRGINIA AREA       640.0   941.692188   634.132033   86.78   
GREAT LAKES AREA                488.0   987.714324   733.521417  126.56   
HOUSTON AREA                    458.0  1127.294127   999.987571   92.38   
LOS ANGELES AREA                671.0  1031.714382   738.616227   96.74   
MIDWEST AREA                    685.0  1095.678701   945.251868   90.42   
NEW ENGLAND AREA         

En `pandas`tenemos una manera más sencilla de mostrar estos resultados

In [53]:
df3[["area","totrev"]].groupby("area").describe()

totrev                                    \
                                count         mean          std     min   
area                                                                      
ATLANTIC SOUTH AREA             592.0   903.497399   697.197572  106.95   
CALIFORNIA NORTH AREA           599.0  1136.348013   922.069148  105.14   
CENTRAL-SOUTH TEXAS AREA        409.0  1174.366577   920.804942   48.53   
CHICAGO AREA                    496.0   892.133206   669.068444   59.90   
DALLAS AREA                     518.0  1258.508784  1267.499852   84.59   
DC-MARYLAND-VIRGINIA AREA       640.0   941.692188   634.132033   86.78   
GREAT LAKES AREA                488.0   987.714324   733.521417  126.56   
HOUSTON AREA                    458.0  1127.294127   999.987571   92.38   
LOS ANGELES AREA                671.0  1031.714382   738.616227   96.74   
MIDWEST AREA                    685.0  1095.678701   945.251868   90.42   
NEW ENGLAND AREA                525.0   941.766038   705.572889  118.51   
NEW YORK CITY AREA             1145.0  1076.263546   869.372842   74.01   
NORTH FLORIDA AREA              442.0   895.489457   642.729028   37.39   
NORTHWEST-ROCKY MOUNTAIN AREA   429.0  1083.179091   977.857935   67.80   
OHIO AREA                       437.0  1006.616499   785.713223  107.12   
PHILADELPHIA AREA               245.0   998.303592   722.348164  133.63   
SOUTH FLORIDA AREA              366.0  1167.546257   891.952255  118.62   
SOUTHWEST AREA                  584.0  1036.548630   914.475459   11.10   
TENNESSEE AREA                  268.0  1042.700597   886.731980  102.03   

                                                                       
                                    25%      50%        75%       max  
area                                                                   
ATLANTIC SOUTH AREA            464.9875  718.370  1135.3050   8171.47  
CALIFORNIA NORTH AREA          535.9900  892.340  1396.4550   7511.11  
CENTRAL-SOUTH TEXAS AREA       594.3300  895.090  1374.5800   5848.39  
CHICAGO AREA                   474.0675  718.855  1086.9250   5710.44  
DALLAS AREA                    581.6975  965.650  1484.8500  13149.85  
DC-MARYLAND-VIRGINIA AREA      525.7200  787.360  1191.5525   5029.68  
GREAT LAKES AREA               503.7325  767.960  1211.8900   5624.71  
HOUSTON AREA                   530.6400  892.180  1443.1400   8719.53  
LOS ANGELES AREA               554.2650  836.770  1271.6150   6120.84  
MIDWEST AREA                   553.0600  854.350  1338.6900  13426.60  
NEW ENGLAND AREA               497.5700  747.130  1185.8000   6939.19  
NEW YORK CITY AREA             562.3400  870.310  1297.9700  10590.01  
NORTH FLORIDA AREA             481.5225  758.190  1134.1950   6009.01  
NORTHWEST-ROCKY MOUNTAIN AREA  524.3700  805.020  1293.2200   7419.73  
OHIO AREA                      512.0700  781.040  1230.5600   8007.95  
PHILADELPHIA AREA              558.7100  849.340  1154.6100   4819.91  
SOUTH FLORIDA AREA             582.0225  914.600  1451.2550   6954.07  
SOUTHWEST AREA                 508.3275  802.025  1187.3275   8772.57  
TENNESSEE AREA                 486.6800  832.400  1282.6150   6751.70

Estos resultados son iguales a los realizados con el ciclo `for`. Los objetos de tipo `DataFrameGroupBy` permiten una rápida salida de generación de estadísticas para grupos de interés.

### Ejercicio 3

Realice una agrupación por la variable creditcd_x y encuentre el valor de la desviación estándar

In [54]:
df3.groupby("creditcd_x").mean()
# realice el ejercicio en este espacio

TypeError: agg function failed [how->mean,dtype->object]

### Etiquetando valores de total de ingresos en categorías

Revisamos el comportamiento de `totrev` y encontramos una gran variabilidad, en ocasiones la agrupación de variables cuantitativas en diferentes categorías nos ayuda a tener una mejor visión del conjunto de datos. 

Crearemos una nueva variable `Nivel_totrev`  que catalogue a la variable total de ingresos `totrev` en 2 posibles niveles, alto en el caso que `totrev`  sea mayor al cuantil 50 y baja en otro caso. cada valor de percentil debe ser calculado por cada `area` para asegurar que el +area geografíca sea evaluada adicionalmente.



In [ ]:
pencentiles50 = df3.groupby("area")["totrev"].quantile(0.5) # percentil 50
print(pencentiles50)

Como nos gustaría etiquetar los ingreos en alto y bajopara  cada área geográfica, haremos uso del método
`np.where()` en la biblioteca numpy. Este método toma una entrada y comprueba una condición lógica: si la
condición es verdadera, devolverá su segundo argumento, mientras que si la condición es falsa, devolverá su
tercer argumento. 


In [ ]:
# Loop a través de los símbolos
areas = ['ATLANTIC SOUTH AREA',
 'CALIFORNIA NORTH AREA',
 'CENTRAL-SOUTH TEXAS AREA',
 'CHICAGO AREA',
 'DALLAS AREA',
 'DC-MARYLAND-VIRGINIA AREA',
 'GREAT LAKES AREA',
 'HOUSTON AREA',
 'LOS ANGELES AREA',
 'MIDWEST AREA',
 'NEW ENGLAND AREA',
 'NEW YORK CITY AREA',
 'NORTH FLORIDA AREA',
 'NORTHWEST-ROCKY MOUNTAIN AREA',
 'OHIO AREA',
 'PHILADELPHIA AREA',
 'SOUTH FLORIDA AREA',
 'SOUTHWEST AREA',
 'TENNESSEE AREA'] # registro de las áreas
lista_df = []
# ciclo sobre todos los símbolos
for i in areas:
    print("Etiqueta por área: " + i)
    temp_df = df3[df3["area"] == i] .copy()
    umbral_punto = pencentiles50.loc[i]
    temp_df["Nivel_totrev"] = np.where(temp_df["totrev"] < umbral_punto, "Bajo", "Alto") 
    lista_df.append(temp_df)
df_con_etiquetas = pd.concat(lista_df)

In [ ]:
df_con_etiquetas.head()

In [ ]:
df_con_etiquetas.head()

Ahora podemos hacer una evaluaciónd de como es el comportamiento de los minutos gastados y el nivel de ingresos de los clientes

### ¿Existe relación alguna entre el total de minutos usados con la compañía y el ingreso total de los clientes?

Para explorar la relación entre el nivel de total de ingresos y el número promedio de llamadas, agrupemos por
Nivel_totrev y miremos ingreso total promedio por cada +area geográfica.



In [ ]:
df_con_etiquetas.groupby(['area','Nivel_totrev'])[['totmou']].mean()

### Ejercicio 4


Escriba el código para categorizar  el total de ingresos  baja, media y alta volatilidad, donde:

`
si totrev > (percentil 75 de totrev para el área dada):
Nivel_totrev = 'Alto'
o si VolStat > (percentil 25 de totrev para el área dada):
Nivel_totrev  = 'Medio'
de lo contrario:
Nivel_totrev  = 'Bajo'`

In [ ]:
pencentiles25 = df3.groupby("area")["totrev"].quantile(0.25) # percentil 50
pencentiles75 = df3.groupby("area")["totrev"].quantile(0.75)

In [ ]:
areas = ['ATLANTIC SOUTH AREA',
 'CALIFORNIA NORTH AREA',
 'CENTRAL-SOUTH TEXAS AREA',
 'CHICAGO AREA',
 'DALLAS AREA',
 'DC-MARYLAND-VIRGINIA AREA',
 'GREAT LAKES AREA',
 'HOUSTON AREA',
 'LOS ANGELES AREA',
 'MIDWEST AREA',
 'NEW ENGLAND AREA',
 'NEW YORK CITY AREA',
 'NORTH FLORIDA AREA',
 'NORTHWEST-ROCKY MOUNTAIN AREA',
 'OHIO AREA',
 'PHILADELPHIA AREA',
 'SOUTH FLORIDA AREA',
 'SOUTHWEST AREA',
 'TENNESSEE AREA'] # registro de las áreas
lista_df = []
# ciclo sobre todos los símbolos
for i in areas:
    print("Etiqueta por área: " + i)
    temp_df = df3[df3["area"] == i] .copy()
    umbral_punto1 = pencentiles25.loc[i]
    umbral_punto2 = pencentiles75.loc[i]
    lv=[]
    for i in temp_df['totrev']:
        if i<umbral_punto1:
            lv.append('Bajo')
        elif i<umbral_punto2:
            lv.append('Medio')
        else:
            lv.append('Alto')
    temp_df['Nivel']=lv
    lista_df.append(temp_df)
df_con_etiquetas = pd.concat(lista_df)
# realice el ejercicio en este espacio

In [ ]:
df_con_etiquetas

In [ ]:
df_con_etiquetas.groupby(['area','Nivel'])[['totmou']].mean()

### Visualización de total de llamadas y total de ingresos

Ya hemos respondido satisfactoriamente a nuestra pregunta original. Sin embargo, no es necesario solamente
analizar los datos en formato tabular. Python contiene una funcionalidad que le permite analizar sus datos
visualmente también.

Usaremos la funcionalidad de pandas sobre la librería estándar de graficación de Python, `matplotlib`. Vamos
a importar la librería e instruir a Jupyter que muestre los gráficos en línea (es decir, mostrar los gráficos en
la pantalla del cuaderno para que podamos verlos mientras ejecutamos el código):



In [ ]:
modules = dir()
print(modules)

In [ ]:
import matplotlib.pyplot as plt
# Graficar en el cuaderno
%matplotlib inline

#### Histograma

In [ ]:
plt.style.use('ggplot')
plt.hist(df3["totrev"],bins=13,color="#237AAF")
plt.title('Histograma de variable totrev')
plt.xlabel("totrev");plt.ylabel("Frecuencia")
plt.show()

In [ ]:
print(plt.style.available)

#### Boxplot

In [ ]:
plt.style.use('ggplot')
plt.boxplot(df3["totrev"])
plt.title('Boxplot de variable totrev')
plt.show()

#### Gráficas a partir de `groupby`

In [ ]:
df4=df_con_etiquetas.groupby(['area','Nivel_totrev'])[['totmou']].mean()
df4.plot(kind='barh',figsize=(10,8))
plt.show()

#### Gráficas de `pandas`  

El modulo`.plot()` genera por defecto una gráfica de línea, existen varios gráficos para los `DataFrameGroup`

    - 'line' : Gráfica de línea
    - 'bar' :  Diagrama de barras vertical
    - 'barh' : Diagrama de barras horizontal
    - 'hist' : histograma
    - 'box' : boxplot
    - 'kde' : Kernel de densidad
    - 'area' : Gráfica de área
    - 'pie' : Diagrama de torta
    - 'scatter' : Diagrama de dispersión
    - 'hexbin' : Diagrama de hexágonos

In [ ]:
df3.columns

In [ ]:
df3["creditcd_x"].value_counts().plot(kind="pie")
plt.show()

Revisando la relación entre las dos variables `totrev` y `totmou`

In [ ]:
df3.plot(kind="scatter",x='totrev',y='totmou',c='blue')

### `Seaborn`

In [ ]:
import seaborn as sns
sns.boxplot(df3["totrev"],orient='v')
plt.title('Boxplot de variable totrev')
plt.show()

### Gráficas de parcela 

In [ ]:
df3

In [ ]:
plt.figure(figsize=(116,4))
sns.pairplot(data=df3, hue='creditcd_x', vars=['totrev','totmou','eqpdays'])
plt.title("Parcelas")
plt.show()

In [ ]:
plt.figure(figsize=(6,4))
sns.boxplot(x='Nivel_totrev',y='totmou',data=df_con_etiquetas)
plt.title('Boxplot múltiple para Nivel de ingresos totales y \n total de minutos gastados')
plt.show()

In [ ]:
sns.catplot(y="totmou", x="Nivel_totrev", kind="box",    data=df_con_etiquetas, height=4, aspect=4, palette='Set2')
sns.catplot(y="totmou", x="Nivel_totrev", kind="violin", data=df_con_etiquetas, height=4, aspect=4, palette='Set2')

In [ ]:
ax = sns.catplot(y="totmou", x="Nivel_totrev", hue='creditcd_x', kind='point',data=df_con_etiquetas, aspect=2)
ax = sns.catplot(x="Nivel_totrev", kind='count',data=df_con_etiquetas, col='creditcd_x')

### Ejercicio 5

Realice un  boxplot para la variable totmou segmentado por las variable  credict_x y Nivel_totrev

In [ ]:
# realice el ejercicio en este espacio

## Resumen


1. Leer datos desde archivos.
2. Agregar y manipular datos .
3. Analizar estadísticas de resumen y reunir información.
4. Usar matplotlib para crear gráficos para hacer análisis visual, además de otras librerías.

Para más información

> [https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html)